### ISSUE TITLE
Why does the input data got echoed in the output?

### ISSUE THREAD

**Issue by knight42** [Tyriar/node-pty#78]
Running the following script

```
const pty = require('node-pty')
const shell = pty.spawn('bash', ['-ri'])
const stdout = process.stdout
shell.on('data', stdout.write.bind(stdout))
process.stdin.on('data', shell.write.bind(shell))
```

I got:

```
[knight@Castle bash.js]$ pwd
pwd
/tmp/bash.js
[knight@Castle bash.js]$
```

Is it possible to turn off the echo?

**Comment by jrop** [comment]
I believe you need this line in there somewhere:
```js
process.stdin.setRawMode(true)
```

**Comment by jrop** [comment]
The reason this is happening is that you have two TTY's:  1) your "master" terminal, and 2) the spawned terminal (`ptw.spawn(...)`).  Each one of these has it's own "echo" logic, and both are echoing when you type.  By setting `process.stdin.setRawMode(true)`, you are telling (1) that you will handle the echo logic*.

*if I understand this correctly.

**Comment by knight42** [comment]
@jrop Thank you so much! After some exploration, I found the solution:

```
const pty = require('node-pty')
const shell = pty.spawn('bash', ['-ri'])
const stdout = process.stdout
const TTY = process.binding('tty_wrap').TTY
const t = new TTY(shell.fd, true)
t.setRawMode(true)
shell.on('data', stdout.write.bind(stdout))
process.stdin.on('data', shell.write.bind(shell))
```

**Comment by knight42** [comment]
@Tyriar I guess `setRawMode` may be exported as a member function of `class Terminal`?

**Comment by jrop** [comment]
@knight42 out of curiosity, did it not work to set `process.stdin.setRawMode(true)` directly (instead of using `tty_wrap`)?

**Comment by knight42** [comment]
@jrop In fact, it works, but not the way I expected.

If I pressed `CTRL-D`, I would get an error:
```
/tmp/bash.js/node_modules/node-pty/lib/unixTerminal.js:77
                throw err;
                ^

Error: This socket is closed
    at PipeSocket.Socket._writeGeneric (net.js:692:19)
    at PipeSocket.Socket._write (net.js:743:8)
    at doWrite (_stream_writable.js:329:12)
    at writeOrBuffer (_stream_writable.js:315:5)
    at PipeSocket.Writable.write (_stream_writable.js:241:11)
    at PipeSocket.Socket.write (net.js:670:40)
    at UnixTerminal.write (/tmp/bash.js/node_modules/node-pty/lib/unixTerminal.js:137:21)
    at emitOne (events.js:96:13)
    at ReadStream.emit (events.js:191:7)
    at readableAddChunk (_stream_readable.js:178:18)
```
besides, I can no longer exit `bash` by `CTRL-C`.

After all, the allocated tty  is  redundant, not my "master" terminal.

**Comment by Tyriar** [comment]
I'd like to get rid of that `tty_wrap` hack if possible as it's modifying the internal node code.

@knight42 maybe to get around that you can switch stdin back when you the terminal gets input?

**Comment by logicminds** [comment]
I am also experiencing this issue.

I recently switched from the older pty module of which this project was forked from and noticed some differences in terms of how writing to stdout and stdin are handled. 

Previously I was using `processPty.stdin.write` to write data to the console.  The cli app (processPty) would read from stdin, process it and then write to stdout.  Because the app wrote to stdout the processPty.on('data') event would be triggered. 

After switching to node-pty I had to refactor the code and use `processPty.write` instead.  I am hoping that this change runs the exact same code as `processPty.stdin.write` used to.  

However, my observations shows that `processPty.write` might be sending to stdin and stdout.  Additionally, my cli app then writes to stdout as well.  So I get double outputs to stdout.

Furthermore, this seems to only occur on the first `processPty.write` as subsequent writes do not exhibit this double stdout write behavior.   Almost as if the initial echo to stdout happens before my cli app has completed initialization/loading (takes about 1500 ms to initialize). 

You can view this intended and actual behavior in my apps below:

 * [original working app](https://demo.puppet-debugger.com/play?content=vars)
 * [refactored with node-pty app](https://puppet-debugger-stage.herokuapp.com/play?content=vars)

This problem looks to be the same as #180 too.

**Comment by corwin-of-amber** [comment]
`shell.fd` is now `shell._fd`. So indeed this should be private data. It still works if you change it; but I haven't found a way to send EOF to the controlled process in this mode.

**Comment by Tyriar** [comment]
@corwin-of-amber fyi https://github.com/microsoft/node-pty/issues/143

**Comment by corwin-of-amber** [comment]
Does sound relevant although the issue is four years old with the last "bump" two years ago.

If anyone is interested in adding new features like this one, they are welcome to do so on the rust-port branch here: https://github.com/corwin-of-amber/node-pty/tree/rust-port
(hopefully will shortly get merged into https://github.com/daniel-brenot/node-pty/tree/rust-port, and if @daniel-brenot manages to convince Microsoft, then perhaps even a branch for it will be created in the official repo.)

**Comment by Tyriar** [comment]
#143 is a very easy change, it's been marked for 4 years with help wanted 🤷. The plan is to maintain a rust branch, but it would be ideal if the APIs were identical. Plus #143 will be a TS only change.

**Comment by daniel-brenot** [comment]
If this is a TS only change, I see no reason to hold up doing this for the rust work. The rust changes most likely wont affect the TS API much, if at all.

**Comment by corwin-of-amber** [comment]
I agree. The only reason I suggested the rust-port branch is because this seems like the only active branch right now. But definitely a PR against https://github.com/daniel-brenot/node-pty/tree/main should work just as well.

**Comment by Tyriar** [comment]
@corwin-of-amber a trivial change like that should get merged quickly, most PRs are limited by me though and sometimes I get behind due to the volume of notifications I get.

**Comment by lafkpages** [comment]
Is there a way to keep the master TTY in TTY mode, and set Node PTY to raw mode? Basically, I want to run commands without echoing back the input commands, shell prompt, etc.

**Comment by daniel-brenot** [comment]
> Does sound relevant although the issue is four years old with the last "bump" two years ago.
> 
> If anyone is interested in adding new features like this one, they are welcome to do so on the rust-port branch here: https://github.com/corwin-of-amber/node-pty/tree/rust-port (hopefully will shortly get merged into https://github.com/daniel-brenot/node-pty/tree/rust-port, and if @daniel-brenot manages to convince Microsoft, then perhaps even a branch for it will be created in the official repo.)

The only issue with the rust branch is that I need help completing it. From what I understand Microsoft doesn't have an issue with merging it in, but ATM I just don't have the ability to get all the functionality working

**Comment by Et7f3** [comment]
> out of curiosity, did it not work to use process.stdin.setRawMode(true) directly

I have a use case where node stdin could be plugged to pipe or pty. This function is added only when stdin is pty. I could do a check if defined but in the case of stdin is a pipe then I would still have the echo. @jrop

> `'\x04'` no longer terminates the input

If you want that you need to return it to cooked mode before sending it. Cooked mode mean those special sequence is interpreted by the kernel driver. It is the default mode when you open a pty with node-pty @corwin-of-amber

> Is there a way to keep the master TTY in TTY mode, and set Node PTY to raw mode? Basically, I want to run commands without echoing back the input commands, shell prompt, etc.

@lafkpages I exprimed the same need in https://github.com/microsoft/node-pty/issues/650. My workaround is to call the stty binary with pty.ptsName getter (thanks for making it public).

`stty -F ${pty.ptsName} raw -echo` with execSync that way once stty give me control back I know my terminal is ready in raw mode. I tested my workaroung by using an helper program that write all 256 chars to stdout and receiving it also as input and echoing. It works pretty good.

A proper solution is to change value given at creation time https://github.com/microsoft/node-pty/blob/d6ce76a801f68730f7204b7b27bb21dbf8e63001/src/unix/pty.cc#L212-L241 I don't have windows at hand but I would call `SetConsoleMode` on `hpc` and removing flag `ENABLE_PROCESSED_OUTPUT`, `ENABLE_VIRTUAL_TERMINAL_PROCESSING` from a value I got from `GetConsoleMode` https://github.com/microsoft/node-pty/blob/d6ce76a801f68730f7204b7b27bb21dbf8e63001/src/win/conpty.cc#L206-L206
